AEP/aep-00004.md

```diff
@@ -0,0 +1,187 @@
+There is no internal retry for static mappings.
+If the mapping is dynamic (eg: with a database), the Load Balancer can retry to create the session on another cluster or raise an error if no cluster is available.
+
+Once a session is bound to a cluster, all requests for this session will raise an error if the cluster becomes unreachable.
+For monitoring requests that span multiple clusters (if allowed), it is not clear how to indicate that one of the clusters is unresponsive, while still returning the response from other clusters.
+The simplest solution would be to silently ignore cluster failures and return the response from all the responsive clusters without any indication of cluster failure.
+
+### Monitoring
+
+The presence of Load Balancers would make the monitoring harder because of the listing across clusters.
+
+The recommended way would be to enforce listing tasks and results on sessions and adapt the GUI to match that requirement.
+To have a global overview, it would be necessary to add some session-wide statistics like the number of tasks completed and in error to enable the view at the session level, without the need to query all the clusters.
+The aggregation could be performed at regular interval on all opened sessions to reduce the number of sessions to query, as a closed session cannot have more tasks added.
+
+### Dynamic clusters
+
+The design impose no constraint on the addition of new clusters during operation.
+It would be a simple matter of allowing a new cluster to be chosen at session creation.
+
+The removal of a cluster is trickier.
+This would be possible only if all the sessions of the cluster are closed to avoid computation loss.
+
+If the monitoring of removed clusters is required, a dedicated database would be needed to store all the metadata of the removed clusters.
+When a RPC targets a removed cluster, this backup database could be requested instead.
+
+## Further considerations
+
+### Global monitoring
+
+Imposing to have a monitoring by session is simple enough, but is not really useful.
+A better approach would be to enable the monitoring spanning all clusters, and allow the listing of the tasks from all clusters.
+
+The main issue for that is the paginated listing of tasks or results.
+Indeed, it is not possible to know what pages of which cluster is required to get the n-th page of the global view.
+
+A way would be to introduce gRPC streaming for listing.
+With that, it would be possible for the Load Balancer to request all the clusters and merge their stream using the specified sort.
+
+It is not clear how this could be integrated into the Web GUI where the pages are mandatory for UX purposes.
+Either the GUI could materialize the whole stream when getting the pages, or a dedicated service could perform this materialization task.
+The latter approach would be more complex especially to avoid system overloads in the presence of many client, but we would entirely compatible with the current API.
```

Comment:
```suggestion
The latter approach would be more complex, especially to avoid system overloads in the presence of many clients, but would be entirely compatible with the current API.
```